power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -21,112 +25,249 @@ namespace detail = power_grid_model::optimizer::detail;
 
 using TrafoGraphIdx = Idx;
 using EdgeWeight = int64_t;
-using WeightedTrafo = std::pair<Idx2D, EdgeWeight>;
-using WeightedTrafoList = std::vector<WeightedTrafo>;
-using RankedTransformerGroups = std::vector<std::vector<Idx2D>>;
 constexpr auto infty = std::numeric_limits<Idx>::max();
 
 struct TrafoGraphVertex {
-    bool is_source{}; // is_source = true if the vertex is a source
+    bool is_source{};
 };
 
 struct TrafoGraphEdge {
-    Idx2D pos{};
+    bool is_trafo{false};
+    Idx2D from_to{};
```

Comment:
Should this be group and pos? ie. idx2d from the container?
The from_to can be found in the edges right?

Reply:
you're right. the current `Idx2D` is specifically for hierarchical indices.

in this case, just `using FromTo = std::pair<Idx, Idx>` would do the trick, or you can create a struct in this namespace that contains `struct FromTo{Idx to{}; Idx from{}};`

Reply:
My initial intent making this change is due to the fact that `Idx2D` is a pain * * * concept to work with outside its own territory. In actuality, the field `is_trafo` and `weight` are redundant. My proposal is to swap the `is_trafo` field to the old `Idx2D id`. Accordingly, the `struct FromTo` as suggested by @mgovers would be probably/eventually `struct FromTo {Idx2D from{}; Idx2D to{}};`. This will be up to discussion per later use in the tap changing optimizer.

Reply:
> But having the RankedTransformerGroups by from/to node and weight would not help in finding the exact regulated object.

Indeed, hence the proposal of swapping the redundant is_trafo to Idx_2D